src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -9561,36 +9561,39 @@ public void M()
     void FalseOut(out string s) => throw null!;
 
 #nullable disable
-    void FalseNOut(out string? ns) => throw null!; // warn 7
+    void FalseNOut(out string? ns) => throw null!; // 9
 }
 ";
             var compilation = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
 
             compilation.VerifyTypes();
             compilation.VerifyDiagnostics(
                 // (52,30): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
-                //     void FalseNOut(out string? ns) => throw null!; // warn 7
+                //     void FalseNOut(out string? ns) => throw null!; // 9
                 Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(52, 30),
                 // (11,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
-                //         s2 = null; // warn 1
+                //         s2 = null; // 1
                 Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "null").WithLocation(11, 14),
                 // (14,9): warning CS8602: Dereference of a possibly null reference.
-                //         ns2 /*T:string?*/ .ToString(); // warn 2
+                //         ns2 /*T:string?*/ .ToString(); // 2
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "ns2").WithLocation(14, 9),
                 // (19,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
-                //         s3 = null; // warn 3
+                //         s3 = null; // 3
                 Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "null").WithLocation(19, 14),
                 // (22,9): warning CS8602: Dereference of a possibly null reference.
-                //         ns3 /*T:string?*/ .ToString(); // warn 4
+                //         ns3 /*T:string?*/ .ToString(); // 4
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "ns3").WithLocation(22, 9),
                 // (27,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
-                //         s4 = null; // warn 5
+                //         s4 = null; // 5
                 Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "null").WithLocation(27, 14),
                 // (30,9): warning CS8602: Dereference of a possibly null reference.
-                //         ns4 /*T:string?*/ .ToString(); // warn 6
+                //         ns4 /*T:string?*/ .ToString(); // 6
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "ns4").WithLocation(30, 9),
+                // (35,14): warning CS8600: Converting null literal or possible null value to non-nullable type.
```

Comment:
>// (35,14): warning CS8600: Converting null literal or possible null value to non-nullable type. [](start = 16, length = 96)

This warning feels unexpected. The goal of the change was to relax the type in some scenarios, but we ended up tightening the type in unrelated scenarios.

Reply:
I left a note above to explain this behavior. This is the same behavior we get from a `var` declaration initialized with a `string~`. The null-state of the value is not-null, so the `var` is a `string!`.
I think this is correct.

---
In reply to: [293147949](https://github.com/dotnet/roslyn/pull/36284#discussion_r293147949) [](ancestors = 293147949)

Reply:
> The null-state of the value is not-null, so the var is a string!. I think this is correct.

I cannot say that I like this behavior.


---
In reply to: [293634127](https://github.com/dotnet/roslyn/pull/36284#discussion_r293634127) [](ancestors = 293634127,293147949)

Reply:
> I cannot say that I like this behavior.

I agree it doesn't feel great, but I don't have a good idea for an alternative that would make sense.

Just to clarify, do you also dislike the behavior in `var x = ...;` scenario?
Let's file an issue if you have a suggestion for better behavior.

---
In reply to: [293894338](https://github.com/dotnet/roslyn/pull/36284#discussion_r293894338) [](ancestors = 293894338,293634127,293147949)